2000-08-05  Zack Weinberg  <zack@wolery.cumb.org>

	* i386.h (FUNCTION_OK_FOR_SIBCALL):  Not OK if DECL's return
	type is a float mode, cfun->decl's return type is not, and
	TARGET_FLOAT_RETURNS_IN_80387.

--- gcc/config/i386/i386.h	2000/07/31 17:33:28	1.121
+++ gcc/config/i386/i386.h	2000/08/05 15:45:43
@@ -1319,10 +1319,16 @@ typedef struct ix86_args {
 
 #define FUNCTION_ARG_PARTIAL_NREGS(CUM, MODE, TYPE, NAMED) 0
 
-/* If PIC, we cannot optimize sibling calls to global functions
-   because the PLT requires %ebx live.  */
-#define FUNCTION_OK_FOR_SIBCALL(DECL) \
-  (DECL && (! flag_pic || ! TREE_PUBLIC (DECL)))
+/* If PIC, we cannot make sibling calls to global functions
+   because the PLT requires %ebx live.
+   If we are returning floats on the register stack, we cannot make
+   sibling calls to functions that return floats.  (The stack adjust
+   instruction will wind up after the sibcall jump, and not be executed.) */
+#define FUNCTION_OK_FOR_SIBCALL(DECL) (DECL \
+   && (! flag_pic || ! TREE_PUBLIC (DECL)) \
+   && (! TARGET_FLOAT_RETURNS_IN_80387 \
+       || ! FLOAT_MODE_P (TYPE_MODE (TREE_TYPE (TREE_TYPE (DECL)))) \
+       || FLOAT_MODE_P (TYPE_MODE (TREE_TYPE (TREE_TYPE (cfun->decl))))))
 
 /* This macro is invoked just before the start of a function.
    It is used here to output code for -fpic that will load the

Mon Jul 31 21:14:49 2000  Jeffrey A Law  (law@cygnus.com)

	* gcc.c-torture/execute/20000731-1.c: New test.

--- gcc/testsuite/gcc.c-torture/execute/20000731-1.c.jj	Thu Aug 17 15:02:49 2000
+++ gcc/testsuite/gcc.c-torture/execute/20000731-1.c	Tue Aug  1 05:14:49 2000
@@ -0,0 +1,24 @@
+double
+foo (void)
+{
+  return 0.0;
+}
+
+void
+do_sibcall (void)
+{
+  (void) foo ();
+}
+
+int
+main (void)
+{
+   double x;
+
+   for (x = 0; x < 20; x++)
+      do_sibcall ();
+   if (!(x >= 10))
+      abort ();
+   exit (0);
+}
+
